[TN-333] Protect Weblogin from automated attack

fajarism opened the pull request:
[[TN-333] Issue](https://github.com/human-internet/humanid-weblogin/issues/65#issuecomment-2813623679)

This PR fixes the TN-333 ticket by adding several features below in the weblogin service:
1. AES encryption, to encrypt the `ip` query params sent to the core service.
2. AES_SECRET_KEY env value.
3. General error message.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced AES-256-GCM encryption for enhanced data security.
  - Added support for encrypting and transmitting the client's IP address during OTP requests.

- **Bug Fixes**
  - Improved error handling and user feedback during OTP login and error scenarios.

- **Chores**
  - Updated configuration to support a new AES secret key.
  - Enforced presence of AES secret key environment variable at startup.
  - Enhanced internal error code management for more robust error processing.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
feat: add encryption helper

Commit:
feat: encrypt requester IP

Lunasocute commented on src/application/config/humanid.php:
```diff
@@ -9,4 +9,5 @@
     'server_id' => $_ENV['HUMANID_SERVER_ID'] ?? '',
     'server_secret' => $_ENV['HUMANID_SERVER_SECRET'] ?? '',
     'protocol' => $_ENV['PROTOCOL'] ?? 'http',
+    'aes_secret_key' => $_ENV['AES_SECRET_KEY'] ?? '',
```
^^^ how about the suggestion?

fajarism replied:
fixed in the latest commit

Lunasocute commented on src/application/controllers/Login.php:
```diff
@@ -30,7 +30,10 @@ public function index()
+            $ip = $this->input->ip_address();
+            $encryptedIp = aes_encrypt_gcm($ip, hex2bin($this->humanid->getAesSecretKey()));
+            $encodedEncryptedIp = urlencode($encryptedIp);
+            $response = $this->humanid->userRequestOTP($dialcode, $phone, $webLoginToken, $this->_app->source, $encodedEncryptedIp, $this->lg->id, $requestId);
```
Will "$this->input->ip_address();" return null under some circumstance?

Commit:
fix: as coderabbit suggested

fajarism commented:
> [Coderabbit](https://github.com/human-internet/humanid-weblogin/pull/67#discussion_r2083526682): Alternatively, if you don't want to throw an exception, implement a validation check before using the key:

Added AES empty validation check in config.php

> [Coderabbit](https://github.com/human-internet/humanid-weblogin/pull/67#discussion_r2083526686): Exposing the AES secret key via a public getter is a security risk
[Coderabbit](https://github.com/human-internet/humanid-weblogin/pull/67#discussion_r2083526684): Extract IP encryption to a reusable method.

Changed method to `encryptIp` instead of `getAesSecretKey`

fajarism commented on src/application/libraries/Humanid.php:
```diff
@@ -30,13 +31,19 @@ function __construct()
+    public function getAesSecretKey()
+    {
+        return $this->aes_secret_key;
+    }
```
fixed in the latest commit

fajarism commented on src/application/controllers/Login.php:
```diff
@@ -338,7 +346,10 @@ public function resend()
+        $ip = $this->input->ip_address();
+        $encryptedIp = aes_encrypt_gcm($ip, hex2bin($this->humanid->getAesSecretKey()));
+        $encodedEncryptedIp = urlencode($encryptedIp);
+        $response = $this->humanid->userRequestOTP($dialcode, $phone, $loginToken, $this->_app->source, $encodedEncryptedIp, $this->lg->id, $requestId);
```
fixed in the latest commit

kofimokome commented on src/application/config/humanid.php:
```diff
@@ -9,4 +9,5 @@
     'server_id' => $_ENV['HUMANID_SERVER_ID'] ?? '',
     'server_secret' => $_ENV['HUMANID_SERVER_SECRET'] ?? '',
     'protocol' => $_ENV['PROTOCOL'] ?? 'http',
+    'aes_secret_key' => $_ENV['AES_SECRET_KEY'] ?? '',
```
For easy setup on local, use the same default key as on the core `ThisIsADefaultSecretPhrase`

fajarism replied:
Hi @kofimokome I left the default key empty because there is an empty env validation function. This was a recommendation from the coderabbit.

kofimokome commented on src/application/helpers/helpers_helper.php:
```diff
@@ -14,3 +14,56 @@ function dd()
+    }
+}
+
+if (!function_exists('aes_decrypt_gcm')) {
+    function aes_decrypt_gcm($encrypted, $key, $aad = '') {
+        $cipher = 'aes-256-gcm';
+        $ivlen = 12;
+        $taglen = 16;
+
+        $data = base64_decode($encrypted);
+        $iv = substr($data, 0, $ivlen);
+        $tag = substr($data, $ivlen, $taglen);
+        $ciphertext = substr($data, $ivlen + $taglen);
+
+        $plaintext = openssl_decrypt(
+            $ciphertext,
+            $cipher,
+            $key,
+            OPENSSL_RAW_DATA,
+            $iv,
+            $tag,
+            $aad
+        );
+
+        return $plaintext;
+    }
+}
```
Since the web login only encrypts the IP. I don't think we need a decrypt function

fajarism replied:
Removed in the latest commit.

Commit:
fix: merge conflict

Commit:
fix: remove decrypt function
